### PR TITLE
feat(sql): add column type validation for ALTER COLUMN statements

### LINKED ISSUE #3515: ALTER COLUMN ADD INDEX on LONG column leads to critical level log message
### Describe the bug

```
2023-06-26T07:21:57.411487Z C i.q.g.e.o.AlterOperation could not alter table [table=my_table, command=4, errno=-1, message=`cannot create index for [column='my_col, type=LONG, path=/var/lib/questdb/db/my_table]`]
```

The log message should have an error level, not a critical one.

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Closes #3515

Also fixes races around `CharSequence` usage in `AlterOperation`. The sequences originated in `SqlCompiler` and were used after the compiler was returned back to the pool leading to data races. The fix is as simple as keeping the compiler around while we execute the statement (see the `OperationExecutor` class).

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 30 / 32 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|5|7|71.43%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 30 / 32 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|5|7|71.43%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|14|14|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/ops/AlterOperationBuilder.java [hunk 1]
```diff
@@ -105,7 +107,7 @@ public void ofAddColumn(
             int indexValueBlockCapacity
     ) {
         assert columnName != null && columnName.length() > 0;
-        extraStrInfo.add(columnName);
+        extraStrInfo.add(Chars.toString(columnName));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/ops/AlterOperationBuilder.java:110]: creating strings might not be necessary here. Tokens coming from tokenized can be made immutable until tokenizer is reset. The latter occurs after SQL compilation is fully done.

```
GenericLexer.immutableOf(tok)
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 30 / 32 (93.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|5|7|71.43%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|14|14|100.00%|

**Review by ideoma — DISMISSED** [review]

File: core/src/main/java/io/questdb/griffin/engine/ops/AlterOperationBuilder.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/ops/AlterOperationBuilder.java]: Marking this one as resolved as the compiler is kept around now: see b2f878f066778a7de84f262eb6779a2a568daa43

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 45 / 50 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|5|7|71.43%|
|:large_blue_circle:|io/questdb/cairo/wal/OperationExecutor.java|13|16|81.25%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|14|14|100.00%|

**Comment by puzpuzpuz** [comment]
@bluestreak01 the PR is all green and should be ready for review.

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**